components/script/stylesheet_loader.rs

```diff
@@ -85,9 +85,21 @@ pub struct StylesheetContext {
     /// The node document for elem when the load was initiated.
     document: Trusted<Document>,
     origin_clean: bool,
+    /// A token which must match the generation id of the `HTMLLinkElement` for it to load the stylesheet.
+    generation_id: GenerationId,
 }
 
-impl PreInvoke for StylesheetContext {}
+impl PreInvoke for StylesheetContext {
+    fn should_invoke(&self) -> bool {
+        // We must first check whether the generations of the context and the element match up,
+        // else we risk applying the wrong stylesheet when responses come out-of-order.
+        let gen = self.elem.root()
```

Comment:
Also, `elem` may be a `<style>` element where there's an `@import`, or the load could be for an `@import` rule in general, in which case the generation id doesn't matter. So we should do the check in `process_response_eof`, an we should check that `StylesheetContextSource` is a `LinkElement`.

Note that the imports can theoretically have the same race if mutated from CSSOM before the load finishes, but we don't expose that right now to the web, so it can't happen, so no need to worry about this in this PR.